PhysicsTools/NanoAOD/plugins/LeptonInJetProducer.cc

```diff
@@ -0,0 +1,285 @@
+    srcEle_(consumes<edm::View<pat::Electron>>(iConfig.getParameter<edm::InputTag>("srcEle"))),
+    srcMu_(consumes<edm::View<pat::Muon>>(iConfig.getParameter<edm::InputTag>("srcMu")))
+  {
+    produces<edm::ValueMap<float>>("lsf3");
+    produces<edm::ValueMap<float>>("lmd3");
+    produces<edm::ValueMap<float>>("lep3pt");
+    produces<edm::ValueMap<float>>("lep3eta");
+    produces<edm::ValueMap<float>>("lep3phi");
+    produces<edm::ValueMap<int>>("lep3id");
+    produces<edm::ValueMap<float>>("lsf3match");
+    produces<edm::ValueMap<int>>("lep3idmatch");
+  }
+  ~LeptonInJetProducer() override {};
+  
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+private:
+  void produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const override ;
+
+  static bool orderPseudoJet(fastjet::PseudoJet j1, fastjet::PseudoJet j2);
+  std::tuple<float,float> calculateLSF(std::vector<fastjet::PseudoJet> iCParticles, std::vector<fastjet::PseudoJet> &ljets,
+				       float ilPt, float ilEta, float ilPhi, int ilId, double dr, int nsj) const;
+
+  edm::EDGetTokenT<pat::PackedCandidateCollection> srcPF_;
+  edm::EDGetTokenT<edm::View<pat::Jet>> srcJet_;
+  edm::EDGetTokenT<edm::View<pat::Electron>> srcEle_;
+  edm::EDGetTokenT<edm::View<pat::Muon>> srcMu_;
+
+};
+
+// ------------ method called to produce the data  ------------
+template <typename T>
+void
+LeptonInJetProducer<T>::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const
+{
+
+    // needs PFcandidate collection (srcPF), jet collection (srcJet), leptons collection 
+    edm::Handle<pat::PackedCandidateCollection> srcPF;
+    iEvent.getByToken(srcPF_, srcPF);
+    edm::Handle<edm::View<pat::Jet>> srcJet;
+    iEvent.getByToken(srcJet_, srcJet);
+    edm::Handle<edm::View<pat::Electron>> srcEle;
+    iEvent.getByToken(srcEle_, srcEle);   
+    edm::Handle<edm::View<pat::Muon>> srcMu;
+    iEvent.getByToken(srcMu_, srcMu);
+
+    unsigned int nJet = srcJet->size();
+    unsigned int nEle = srcEle->size();
+    unsigned int nMu  = srcMu->size();
+
+    std::vector<float> *vlsf3 = new std::vector<float>;
+    std::vector<float> *vlmd3 = new std::vector<float>;
+    std::vector<float> *vlep3pt = new std::vector<float>;
+    std::vector<float> *vlep3eta = new std::vector<float>;
+    std::vector<float> *vlep3phi = new std::vector<float>;
+    std::vector<int> *vlep3id = new std::vector<int>;
+
+    std::vector<float> *vlsf3match = new std::vector<float>;
+    std::vector<int> *vlep3idmatch = new std::vector<int>;
+
+    // Matching electron and pfcands  
+    const pat::PackedCandidate *ele_pfmatch = nullptr;
+    double pfDR = 0.05;
+    for (unsigned int il(0); il < nEle; il++) {
+      auto itLep = srcEle->ptrAt(il);
+      for (const pat::PackedCandidate &itPF : *srcPF) {
+	double dR = reco::deltaR(itLep->eta(), itLep->phi(), itPF.eta(), itPF.phi());
```

Comment:
matching by deltaR is not needed, please use matching by ref (see MiniAOD workbook for navigation details)